File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// so if we're at any given node, do determine the current depth would be the max between
// the two children, and then you want to return 1 + that to account for the node itself

class Solution 
{
public:
    int maxDepth(TreeNode* root) 
    {
        // a null node contributes nothing
        if (!root) { return 0; }

        return 1 + std::max(maxDepth(root->left), maxDepth(root->right));
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node) -> int:
            nonlocal res
            
            if not node:
                return 0

            # at a given node the max it can contribute is either both its children
            left_height = dfs(node.left)
            right_height = dfs(node.right)

            res = max(res, left_height + right_height)
            return 1 + max(left_height, right_height)
            
        dfs(root)
        return res